### PR TITLE
Enhance lookup for IdTokens/AppMetadata

### PR DESCRIPTION
**Pull request by sameerag**
This PR:

* adds `filter` based lookup for `idTokens` and `appMetadata` (in lieu of respecting environment aliases)
* Checks for `undefined` objects while retrieving in cache

### CODE REVIEW

**Review by hectormmg — APPROVED** [review]
LGTM!

File: lib/msal-node/src/cache/Storage.ts [hunk 1]
```diff
@@ -194,8 +194,9 @@ export class Storage extends CacheManager {
      */
     getIdTokenCredential(idTokenKey: string): IdTokenEntity | null {
         const credType = CredentialEntity.getCredentialType(idTokenKey);
-        if (credType === CredentialType.ID_TOKEN) {
-            return this.getItem(idTokenKey) as IdTokenEntity;
+        const idToken = this.getItem(idTokenKey);
+        if (idToken && credType === CredentialType.ID_TOKEN) {
```
Comment (tnorling) [inline lib/msal-node/src/cache/Storage.ts:198]: Don't we have `isIdTokenEntity` we can use here instead?

File: lib/msal-node/src/cache/Storage.ts [hunk 2]
```diff
@@ -215,8 +216,9 @@ export class Storage extends CacheManager {
      */
     getAccessTokenCredential(accessTokenKey: string): AccessTokenEntity | null {
         const credType = CredentialEntity.getCredentialType(accessTokenKey);
-        if (credType === CredentialType.ACCESS_TOKEN) {
-            return this.getItem(accessTokenKey) as AccessTokenEntity;
+        const accessToken = this.getItem(accessTokenKey);
+        if (accessToken && credType === CredentialType.ACCESS_TOKEN) {
```
Comment (tnorling) [inline lib/msal-node/src/cache/Storage.ts:220]: Same q for access token and refresh token

File: lib/msal-node/src/cache/Storage.ts (hunk 1 above)
Comment (sameerag) [reply lib/msal-node/src/cache/Storage.ts:198]: @pkanher617  and me plan to add the validation in a single PR for browser and node, hence did not do it here. But yes, plans are there to add it. Currently this change is needed as `null` or `undefined` items are being accessed when we do all key sweep and failing `silent` calls for `msal-node`

File: lib/msal-node/src/cache/Storage.ts (hunk 2 above)
Comment (sameerag) [reply lib/msal-node/src/cache/Storage.ts:220]: Answered above.

File: lib/msal-node/src/cache/Storage.ts (hunk 1 above)
Comment (pkanher617) [reply lib/msal-node/src/cache/Storage.ts:198]: validation was already added for browser, it might be good to add it here for node now.

Comment (sameerag) [reply lib/msal-node/src/cache/Storage.ts:198]: :-D Didn't know you already did; As discussed, will be adding this soon and some more tests.

Comment (sameerag) [reply lib/msal-node/src/cache/Storage.ts]: Made changes.

File: lib/msal-node/src/cache/Storage.ts (hunk 2 above)
Comment (sameerag) [reply lib/msal-node/src/cache/Storage.ts]: Done.

File: lib/msal-node/src/cache/Storage.ts [hunk 3]
```diff
@@ -194,8 +194,9 @@ export class Storage extends CacheManager {
      */
     getIdTokenCredential(idTokenKey: string): IdTokenEntity | null {
         const credType = CredentialEntity.getCredentialType(idTokenKey);
-        if (credType === CredentialType.ID_TOKEN) {
-            return this.getItem(idTokenKey) as IdTokenEntity;
+        const idToken = this.getItem(idTokenKey) as IdTokenEntity;
+        if (IdTokenEntity.isIdTokenEntity(idToken) && credType === CredentialType.ID_TOKEN) {
```
Comment (tnorling) [inline lib/msal-node/src/cache/Storage.ts:198]: credType check is part of validator
```suggestion
        if (idToken && IdTokenEntity.isIdTokenEntity(idToken)) {
```

File: lib/msal-node/src/cache/Storage.ts [hunk 4]
```diff
@@ -215,8 +216,9 @@ export class Storage extends CacheManager {
      */
     getAccessTokenCredential(accessTokenKey: string): AccessTokenEntity | null {
         const credType = CredentialEntity.getCredentialType(accessTokenKey);
-        if (credType === CredentialType.ACCESS_TOKEN) {
-            return this.getItem(accessTokenKey) as AccessTokenEntity;
+        const accessToken = this.getItem(accessTokenKey) as AccessTokenEntity;
+        if (AccessTokenEntity.isAccessTokenEntity(accessToken) && credType === CredentialType.ACCESS_TOKEN) {
```
Comment (tnorling) [inline lib/msal-node/src/cache/Storage.ts:220]:
```suggestion
        if (accessToken && AccessTokenEntity.isAccessTokenEntity(accessToken)) {
```

File: lib/msal-node/src/cache/Storage.ts [hunk 5]
```diff
@@ -236,8 +238,9 @@ export class Storage extends CacheManager {
      */
     getRefreshTokenCredential(refreshTokenKey: string): RefreshTokenEntity | null {
         const credType = CredentialEntity.getCredentialType(refreshTokenKey);
-        if (credType === CredentialType.REFRESH_TOKEN) {
-            return this.getItem(refreshTokenKey) as RefreshTokenEntity;
+        const refreshToken = this.getItem(refreshTokenKey) as RefreshTokenEntity;
+        if (RefreshTokenEntity.isRefreshTokenEntity(refreshToken) && credType === CredentialType.REFRESH_TOKEN) {
```
Comment (tnorling) [inline lib/msal-node/src/cache/Storage.ts:242]:
```suggestion
        if (refreshToken && RefreshTokenEntity.isRefreshTokenEntity(refreshToken)) {
```

File: lib/msal-node/src/cache/Storage.ts (hunk 3 above)
Comment (sameerag) [reply lib/msal-node/src/cache/Storage.ts:198]: Actually we wouldn't need `idToken &&` at all as `IdTokenEntity.isIdTokenEntity` takes care of an undefined entity. Let me make those changes. And yes, `CredentialType` check is also covered there.

**Review by tnorling — APPROVED** [review]
LGTM

**Pull request merged**